server/task_executor_pitr_cutover.go

```diff
@@ -138,3 +137,115 @@ func (exec *PITRCutoverTaskExecutor) pitrCutover(ctx context.Context, task *api.
+		return nil, "", err
+	}
+	payload := api.BackupPayload{BinlogInfo: binlogInfo}
+	payloadBytes, err := json.Marshal(payload)
+	if err != nil {
+		return nil, "", err
+	}
+
+	log.Debug("Starting new transaction and unlock tables")
+	options := sql.TxOptions{ReadOnly: true}
+	// Beginning a transaction in the same session will implicitly release existing table locks.
+	// ref: https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html, section "Interaction of Table Locking and Transactions".
+	txn, err := conn.BeginTx(ctx, &options)
+	if err != nil {
+		return nil, "", err
+	}
+
+	return txn, string(payloadBytes), nil
+}
+
+func (exec *PITRCutoverTaskExecutor) backupDatabaseAfterPITR(ctx context.Context, conn *sql.Conn, txn *sql.Tx, backupPayload string, driver db.Driver, store *store.Store, database *api.Database, dataDir string, suffixTs int64) error {
+	backupName := fmt.Sprintf("%s-%s-pitr_%d", api.ProjectShortSlug(database.Project), api.EnvSlug(database.Instance.Environment), suffixTs)
+	if err := createBackupDirectory(dataDir, database.ID); err != nil {
+		return fmt.Errorf("failed to create backup directory, error: %w", err)
+	}
+	path := getBackupRelativeFilePath(database.ID, backupName)
+	migrationHistoryVersion, err := getLatestSchemaVersion(ctx, driver, database.Name)
+	if err != nil {
+		return fmt.Errorf("failed to get migration history for database %q, error: %w", database.Name, err)
+	}
+	backupCreate := &api.BackupCreate{
+		CreatorID:               api.SystemBotID,
+		DatabaseID:              database.ID,
+		Name:                    backupName,
+		Type:                    api.BackupTypeAutomaticPITRCutover,
+		MigrationHistoryVersion: migrationHistoryVersion,
+		StorageBackend:          api.BackupStorageBackendLocal,
+		Path:                    path,
+	}
+	backup, err := store.CreateBackup(ctx, backupCreate)
+	if err != nil {
+		if common.ErrorCode(err) == common.Conflict {
+			return fmt.Errorf("backup name %q already exists", backupName)
+		}
+		return fmt.Errorf("failed to create backup %q, error: %w", backupName, err)
+	}
+
+	backupPatch := api.BackupPatch{
+		ID:        backup.ID,
+		Status:    string(api.BackupStatusDone),
+		UpdaterID: api.SystemBotID,
+		Comment:   "",
+		Payload:   backupPayload,
+	}
+	backupFile, err := os.Create(filepath.Join(dataDir, path))
+	if err != nil {
+		return fmt.Errorf("failed to create backup path %q", path)
+	}
+	binlogDir := getBinlogAbsDir(dataDir, database.Instance.ID)
+	if err := createBinlogDir(dataDir, database.Instance.ID); err != nil {
+		return err
+	}
+	mysqlDriver, ok := driver.(*mysql.Driver)
+	if !ok {
+		log.Error("failed to cast driver to mysql.Driver")
+		return fmt.Errorf("[internal] cast driver to mysql.Driver failed")
+	}
+	mysqlDriver.SetUpForPITR(exec.mysqlutil, binlogDir)
+
+	// Asynchronously dump the database and update the backup state.
+	go func() {
+		log.Debug("Taking a backup with type AUTOMATIC_PITR_CUTOVER asynchronously")
+		defer conn.Close()
```

Comment:
We use `defer conn.Close()` here. If we encounter some error before the line#226, the conn will leak?